### DISCUSSION TITLE
Consider redesigning ClientWebSocketResponse.receive_* methods

### DISCUSSION THREAD

**Discussion by x0day** [aio-libs/aiohttp#9474] [Ideas]
Now the `ClientWebSocketResponse.receive` method will return a `WS_CLOSED_MESSAGE` when the session is closing, and then a TypeError will be raised in the receive_* method. This is a very unfriendly design, and we cannot capture the unexpected closing situation well.

In addition, the closed check also has a lag, for example, the following code does not work properly.

```python
if not session.closed:
    msg = session.receive()
```

It must be like this:

```python
msg = session.receive()
    if session.closed:
        ...
```

**Comment by Dreamsorcerer** [comment]
> ```python
> if not session.closed:
>     msg = session.receive()
> ```

Shouldn't that be an `await session.receive()`? In which case, isn't that obvious that the session could close while it is waiting for a message?

As for the TypeError, that has discussions elsewhere in other issues and may be improved upon in a future release.
